src/storage.rs

```diff
@@ -14,99 +16,333 @@ pub trait StorageBase {
 
 /// Typed component storage trait.
 pub trait Storage<T>: StorageBase + Sized {
+    /// Used during iterator
+    type UnprotectedStorage: UnprotectedStorage<T>;
+
     /// Create a new storage. This is called when you register a new
     /// component type within the world.
     fn new() -> Self;
+    /// Insert a new data for a given entity.
+    fn insert(&mut self, Entity, T);
     /// Try reading the data associated with an entity.
     fn get(&self, Entity) -> Option<&T>;
     /// Try mutating the data associated with an entity.
     fn get_mut(&mut self, Entity) -> Option<&mut T>;
-    /// Insert a new data for a given entity.
-    fn insert(&mut self, Entity, T);
     /// Remove the data associated with an entity.
     fn remove(&mut self, Entity) -> Option<T>;
+    /// splits the bitset from the storage for use
+    /// by the join iterator.
+    fn entities(&self) -> (&BitSet, &Self::UnprotectedStorage);
```

Comment:
Perhaps, we could find a better name for this method. `entities` here returns neither entities container, or an iterator.
Perhaps `open` and `open_mut`?